File: Ruby/day03.rb
#!/usr/bin/env ruby

require 'util'
require 'geometry'
require 'grid'

# INPUT_FILE = 'day03_test.txt'
INPUT_FILE = 'day03_challenge.txt'

input = Util.read_input("../Input/#{INPUT_FILE}")

puts 'Advent of Code 2023, Day 03: Gear Ratios'

def solve_part_one(grid, lookup)
  valid_part_nums = []
  
  grid.coords.each do |coord|
    val = grid.get_s(coord)
    next if val =~ /\d/
    sym_ext = Extent.new(coord).inset(-1)
    # puts "Extent for symbol #{val} is #{sym_ext.to_s}"
    lookup.keys.each do |pn_ext|
      pn = lookup[pn_ext]
      if sym_ext.intersect(pn_ext) then
        valid_part_nums.push(pn)
      end
    end
  end
  
  sum = valid_part_nums.inject(:+)
  puts "The sum of all valid part numbers is #{sum}"
end


def solve_part_two(grid, lookup)
  gear_ratios = []
  
  grid.coords.each do |coord|
    val = grid.get_s(coord)
    next if val != '*'
    gear_ext = Extent.new(coord).inset(-1)
    part_nums = []
    lookup.keys.each do |ext|
      part_num = lookup[ext]
      if gear_ext.intersect(ext) then
        part_nums.push(part_num)
      end
    end
    if part_nums.length == 2 then
      ratio = part_nums[0] * part_nums[1]
      gear_ratios.push(ratio)
    end
  end
  
  sum = gear_ratios.inject(:+)
  puts "The sum of all gear ratios is #{sum}"
end

def find_part_numbers(grid)
  result = {}
  
  part_num = ''
  start_coord = end_coord = nil
  
  # Coords are returned in "reading order". Good for parsing the part numbers.
  grid.extent.all_coords.each do |coord|
    val = grid.get_s(coord)
    is_num = val =~ /\d/
    if is_num then
      part_num += val
      if start_coord == nil then
        start_coord = coord
      end
      end_coord = coord
    else
      if start_coord != nil then
        ext = Extent.new(start_coord, end_coord)
        # Using the extent as the key because part numbers are duplicated
        result[ext] = Integer(part_num)
        start_coord = end_coord = nil
        part_num = ''
      end
    end
  end  
  result
end

blueprint = Grid.new('.', :QUEEN)
blueprint.load(input)
part_number_extents = find_part_numbers(blueprint)
# part_number_extents.keys.each do |key|
#   puts "#{part_number_extents[key]} => #{key.to_s}"
# end

solve_part_one blueprint, part_number_extents 
solve_part_two blueprint, part_number_extents 

File: PHP/src/day03.php
<?php

namespace sjb\aoc2023;

require 'lib/Util.php';
require 'lib/Geometry.php';
require 'lib/Grid.php';

echo "Advent of Code 03, Day 03: Gear Ratios\n";

$INPUT_DIR = '../Input/';
// $INPUT_FILE = 'day03_test.txt';
$INPUT_FILE = 'day03_challenge.txt';
$INPUT_INDEX = 0;

$input = read_grouped_input($INPUT_DIR . $INPUT_FILE, $INPUT_INDEX);

$schematic = new Grid2D('QUEEN', '.');
$schematic->load($input);
// $schematic->print();

$partsNextToGears = []; // Side effect of part 1, ready for part 2

$result1 = solvePartOne($schematic);
echo "Part One: The sum of the part numbers is $result1\n";

$result2 = solvePartTwo($partsNextToGears);
echo "Part Two: The sum of gear ratios is $result2\n";


function solvePartOne(Grid2D $schematic):int {
	$num = '';
	$start = null;
	$numbers = [];
	$ext = $schematic->getExtent()->inset(-1);
	
	foreach ($ext->getAllCoords() as $coord) { // Is in "reading order"
		$val = $schematic->getValue($coord);
		if (is_numeric($val)) {
			if (is_null($start)) { $start = $coord; }
			$num .= $val;
		}
		else if (strlen($num) > 0) {
			$searchExt = new Extent2D($start, $coord->offset('<')); // Rewind to previous coord
			$symLocation = findAdjacentSymbol($searchExt, $schematic);
			if ($symLocation) {
				array_push($numbers, intval($num));
				
				// Storing info for part 2
				global $partsNextToGears;
				if ($schematic->getValue($symLocation) == '*') {
					$key = strval($symLocation);
					if (!array_key_exists($key, $partsNextToGears)) { $partsNextToGears[$key] = []; }
					array_push($partsNextToGears[$key], intval($num));
				}
			}
			$num = '';
			$start = null;
		}
	}
	
	$sum = array_reduce($numbers, fn($sum,$num) => $sum + $num, 0);
	return $sum;
}

function solvePartTwo(array $input):int  {
	$sum = 0;
	foreach ($input as $coord => $numbers) {
		if (count($numbers) == 2) {
			$ratio = $numbers[0] * $numbers[1];
			$sum += $ratio;
		}
	}
	return $sum;
}

function findAdjacentSymbol(Extent2D $extent, Grid2D $schematic): ?Coord2D {
	$outset = $extent->inset(-1);
	foreach ($outset->getAllCoords() as $coord) {
		$val = $schematic->getValue($coord);
		if ($val != $schematic->getDefault() && !is_numeric($val)) {
			return $coord;
		}
	}
	return null;
}